src/reactor.rs

```diff
@@ -0,0 +1,741 @@
+                    .flatten()
+                    .map(|k| k.into_owned()),
+            },
+
+            Request::Initialize {
+                id,
+                start,
+                validity,
+            } => Response::Initialize {
+                id,
+                initialized: keychain
+                    .public(start)
+                    .ok()
+                    .flatten()
+                    .map(|public| self.add_validator(public.into_owned(), start, validity))
+                    .is_some(),
+            },
+
+            Request::Round { id } => {
+                let height = self.height();
+                let round = self.round(moderator.now());
+                let leader = self.leader(round).copied().unwrap_or_default();
+                let public = keychain
+                    .public(height)
+                    .unwrap_or_default()
+                    .map(|p| p.into_owned())
+                    .unwrap_or_default();
+                let step = self.validator_step(height, round, &public);
+
+                Response::Round {
+                    id,
+                    height,
+                    round,
+                    leader,
+                    step,
+                }
+            }
+        };
+
+        let response = Message::Response(response);
+
+        moderator.send(response, self.timeout).await;
+    }
+
+    /// Receive a new message, mutating the internal state
+    pub async fn receive<K, M>(&mut self, keychain: &K, moderator: &mut M, message: Message)
+    where
+        K: Keychain,
+        M: Moderator,
+    {
+        #[cfg(feature = "trace")]
+        tracing::trace!("receiving message {:?}", message);
+
+        match message {
+            Message::Event(e) => self.receive_event(e).await,
+            Message::Notification(n) => self.receive_notification(keychain, moderator, n).await,
+            Message::Request(r) => self.receive_request(keychain, moderator, r).await,
+            Message::Response(_) => (),
+        }
+    }
+
+    /// Check the current status of the reactor, producing an event, if applicable
+    pub async fn heartbeat<K, M>(&mut self, keychain: &K, moderator: &mut M) -> Result<(), Error>
+    where
+        K: Keychain,
+        M: Moderator,
+    {
+        let height = self.height();
+
+        #[cfg(feature = "trace")]
+        tracing::trace!("heartbeat height {}", height);
+
+        // If no public key is available in the keychain, the node is idle
+        let public = match keychain
+            .public(height)
+            .map_err(|_| Error::ResourceNotAvailable)?
+        {
+            Some(p) => p,
+            None => {
+                let idle = Message::Event(Event::Idle);
+
+                // FIXME maybe clean all queues?
+                moderator.send(idle, self.timeout).await;
+
+                // Lock the thread until next inbound notification
+                if let Ok(Some(m)) = moderator.inbound_blocking() {
```

Comment:
This seems suspicous to me.. Could this lead to a deadlock if we set the tokio runtime to single-threaded mode?

Reply:
Inside `TokioModerator`, it is calling `blocking_recv` that can "This function panics if called within an asynchronous execution context." https://docs.rs/tokio/0.3.5/tokio/sync/mpsc/struct.Receiver.html#method.blocking_recv

Reply:
We can't proceed with consensus voting if we don't have the identity of the current node

Reply:
This needs to be removed once we implement #16 as we will no longer require a key